### PR TITLE
all: improve EstimateGas API 

### PR DESCRIPTION
**Pull request by JukLee0ira**
# Proposed changes
this PR separate consensus errors and EVM Internal errors
### Test
test contract is following:
```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

contract TestRevert {
    function revertLogic(address payable recipient)
        external
        payable
        returns (uint256)
    {
        recipient.transfer(msg.value);
        require(msg.value > 10, "Transfer amount must be greater than ten");

        return msg.value;
    }

    // Function to allow contract to send ETH to another account
    function sendXDC(address payable recipient, uint256 amount) external returns (uint256){
        // Transfer the specified amount to the recipient
        recipient.transfer(amount* 1 ether);

        return amount;
        
    }

    receive() external payable {
    }

    fallback() external payable {
    }
}


```
#### Gas Consumption Comparison
**eth_estimateGas**
The `curl` command I used:
- address with balance  : `0x873C36f9Fd02e0C57a393aFE80D14f244fE04378`
- address without balance: `0xFb51679CeF4CF53152B531e6E437C727aDfF533e`
```bash
RPC="http://127.0.0.1:8545"
curl -s -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc": "2.0",
  "id": 2,
  "method": "eth_estimateGas",
  "params": [
    {
      "from": <wallet address>,  
      "to": "0xEb14F05e78bDB995a55aAc7D2b2D26D2215F22Fd",
      "value":"0xc8",
      "data": "0x8a73d3ab000000000000000000000000eb14f05e78bdb995a55aac7d2b2d26d2215f22fd"
    },
    "latest"
  ]}' 
```
and their response:

| Balance | PR are applied | Response                                      |
| ------- | -------------- | --------------------------------------------- |
| Yes     | Yes            | {"jsonrpc":"2.0","id":1002,"result":"0x8066"} |
| Yes     | No             | {"jsonrpc":"2.0","id":1002,"result":"0x8066"} |
| No      | Yes            | {"jsonrpc":"2.0","id":1002,"result":"0x8066"} |
| No      | No             | {"jsonrpc":"2.0","id":1002,"result":"0x8066"} |

---
**eth_call**
The `curl` command I used:
```bash
RPC="http://127.0.0.1:8545"
curl -s -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc": "2.0",
  "id": 2,
  "method": "eth_call",
  "params": [
    {
      "from": "0x873C36f9Fd02e0C57a393aFE80D14f244fE04378",  
      "to": "0xEb14F05e78bDB995a55aAc7D2b2D26D2215F22Fd",
      "data": "0x1afd35fe000000000000000000000000885c1e1b9c24758b56b6a36c13a94efdb4e4e3b10000000000000000000000000000000000000000000000000000000000000001"
    },
    "latest"
  ]}'
  ```
and their response:

| PR are applied | Response                                                                                                      |
| -------------- | ------------------------------------------------------------------------------------------------------------- |
| Yes            | {"jsonrpc":"2.0","id":1002,"result":"0x0000000000000000000000000000000000000000000000000000000000000001"}<br> |
| No             | {"jsonrpc":"2.0","id":1002,"result":"0x0000000000000000000000000000000000000000000000000000000000000001"}<br> |

##### Request（The balance is 100, and an internal transaction attempts to transfer 200 to the address`0x873C36f9Fd02e0C57a393aFE80D14f244fE04378`）
```bash
RPC="http://127.0.0.1:8545"
curl -s -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc": "2.0",
  "id": 0,
  "method": "eth_call",
  "params": [
    {            
      "from": "0x873C36f9Fd02e0C57a393aFE80D14f244fE04378",
      "to": "0xEb14F05e78bDB995a55aAc7D2b2D26D2215F22Fd",
      "data": "0x1afd35fe000000000000000000000000885c1e1b9c24758b56b6a36c13a94efdb4e4e3b100000000000000000000000000000000000000000000000000000000000000c8",
    },
  ]}' | jq
```
##### Response
```bash
{
    "jsonrpc": "2.0",
    "id": 1002,
    "error": {
        "code": -32000,
        "message": "execution reverted"
    }
}
```
##### Request（gas exceeds the gasLimit）
```bash
RPC="http://127.0.0.1:8545"
curl -s -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc": "2.0",
  "id": 0,
  "method": "eth_call",
  "params": [
    {            
      "from": "0x873C36f9Fd02e0C57a393aFE80D14f244fE04378",
      "to": "0xEb14F05e78bDB995a55aAc7D2b2D26D2215F22Fd",
      "gas": "0x10",
      "data": "0x1afd35fe000000000000000000000000885c1e1b9c24758b56b6a36c13a94efdb4e4e3b10000000000000000000000000000000000000000000000000000000000000008",
    },
  ]}' | jq
```
##### Response
```bash
{
    "jsonrpc": "2.0",
    "id": 1002,
    "error": {
        "code": -32000,
        "message": "err: intrinsic gas too low: have 16, want 22680 (supplied gas 16)"
    }
}
```
##### Request(revert caused by business logic)
```bash
RPC="http://127.0.0.1:8545"
curl -s -X POST -H "Content-Type: application/json" ${RPC} -d '{
  "jsonrpc": "2.0",
  "id": 0,
  "method": "eth_estimateGas",
    "params": [
        {
		"from": "0x873C36f9Fd02e0C57a393aFE80D14f244fE04378",
		"to": "0xEb14F05e78bDB995a55aAc7D2b2D26D2215F22Fd",
		"value":"0x3",
		"data": "0x8a73d3ab000000000000000000000000eb14f05e78bdb995a55aac7d2b2d26d2215f22fd"
        }
    ]}' | jq
```

##### Response
```bash
{
    "jsonrpc": "2.0",
    "id": 1002,
    "error": {
        "code": -32000,
        "message": "always failing transaction (execution reverted) (Transfer amount must be greater than ten)"
    }
}
```
## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement
- [x] N/A
## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [ ] Geth
- [ ] Smart Contract
- [ ] External components
- [x] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [ ] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [x] N/A

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement
- [x] N/A
## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [ ] Geth
- [ ] Smart Contract
- [ ] External components
- [x] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [ ] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [x] N/A

## Types of changes

What types of changes does your code introduce to XDC network?
_Put an `✅` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Regular KTLO or any of the maintaince work. e.g code style
- [ ] CICD Improvement
- [x] N/A
## Impacted Components
Which part of the codebase this PR will touch base on,

_Put an `✅` in the boxes that apply_

- [ ] Consensus
- [ ] Account
- [ ] Network
- [ ] Geth
- [ ] Smart Contract
- [ ] External components
- [x] Not sure (Please specify below)

## Checklist
_Put an `✅` in the boxes once you have confirmed below actions (or provide reasons on not doing so) that_

- [ ] This PR has sufficient test coverage (unit/integration test) OR I have provided reason in the PR description for not having test coverage
- [ ] Provide an end-to-end test plan in the PR description on how to manually test it on the devnet/testnet.
- [ ] Tested the backwards compatibility.
- [ ] Tested with XDC nodes running this version co-exist with those running the previous version.
- [ ] Relevant documentation has been updated as part of this PR
- [x] N/A

### CODE REVIEW

File: core/vm/logger.go [hunk 1]
```diff
@@ -32,6 +33,8 @@ import (
 	"github.com/holiman/uint256"
 )
 
+var errTraceLimitReached = errors.New("the number of logs reached the specified limit")
```
Comment (gzliudan) [inline core/vm/logger.go:36]: `errTraceLimitReached` is not used

File: core/vm/instructions.go [hunk 2]
```diff
@@ -17,15 +17,22 @@
...
 	"github.com/XinFinOrg/XDPoSChain/core/types"
 	"github.com/XinFinOrg/XDPoSChain/params"
 	"github.com/holiman/uint256"
 	"golang.org/x/crypto/sha3"
 )
 
+var (
+	bigZero = new(big.Int)
+	tt255   = math.BigPow(2, 255)
+)
```
Comment (gzliudan) [inline core/vm/instructions.go:34]: `bigZero` and  `tt255` are not used

File: accounts/abi/bind/backends/simulated.go [hunk 3]
```diff
@@ -160,6 +160,12 @@ func NewSimulatedBackend(alloc core.GenesisAlloc) *SimulatedBackend {
 	return backend
 }
 
+// Close terminates the underlying blockchain's update loop.
+func (b *SimulatedBackend) Close() error {
+	b.blockchain.Stop()
+	return nil
+}
+
```
Comment (gzliudan) [inline accounts/abi/bind/backends/simulated.go:168]: It seems that the function `Close()` is not used.

File: core/error.go [hunk 4]
```diff
@@ -54,3 +82,11 @@ var (
 	// ErrGasUintOverflow is returned when calculating gas usage.
 	ErrGasUintOverflow = errors.New("gas uint64 overflow")
 )
+
+type ConsensusError struct {
+	Reason error
+}
+
```
Comment (gzliudan) [inline core/error.go:89]: Where is ConsensusError used ?

**Comment by gzliudan** [comment]
Can you update the function `UnpackRevert` to latest version:

https://github.com/ethereum/go-ethereum/blob/master/accounts/abi/abi.go#L275-L311 ?

**Comment by JukLee0ira** [comment]
Revert Logic Test have been added.

**Comment by gzliudan** [comment]
missing file accounts/abi/bind/backends/simulated_test.go

File: core/state_transition.go [hunk 5]
```diff
@@ -317,16 +344,16 @@ func (st *StateTransition) TransitionDb(owner common.Address) (ret []byte, usedG
 	// Check clauses 4-5, subtract intrinsic gas if everything is correct
 	gas, err := IntrinsicGas(st.data, st.msg.AccessList(), contractCreation, homestead, rules.IsEIP1559)
 	if err != nil {
-		return nil, 0, false, err, nil
+		return nil, err
 	}
 	if st.gas < gas {
-		return nil, 0, false, fmt.Errorf("%w: have %d, want %d", ErrIntrinsicGas, st.gas, gas), nil
+		return nil, ErrIntrinsicGas
```
Comment (gzliudan) [inline core/state_transition.go:350]: This line should be:

```go
return nil, fmt.Errorf("%w: have %d, want %d", ErrIntrinsicGas, st.gas, gas)
```

**Comment by gzliudan** [comment]
The line `recipient.transfer(msg.value);` in your contract is never reverted. Please modify your test contract.

**Review by gzliudan — APPROVED** [review]

**Pull request merged**